[inductor][scheduler] Use set for origin

anijain2305 opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #119861

xref - https://github.com/pytorch/pytorch/issues/119440

This avoids node > node comparison if the origin order is same in the origins tuple. However, I am unable to come up with a test case where this could happen.

cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @peterbell10 @ipiszy @yf225 @chenyang78 @kadeng @muchulee8 @aakhundov @ColinPeppler

anijain2305 committed:
[inductor][scheduler] Use set for origin

[ghstack-poisoned]

Skylion007 commented on torch/_inductor/scheduler.py:
```diff
@@ -2248,9 +2248,9 @@ def get_order(n):
+        origins = {(get_order(e), e) for n in node.get_nodes() for e in n.node.origins}
         if origins:
-            _, last = max(origins)
+            _, last = max(origins, key=lambda x: x[0])
```
```suggestion
            _, last = max(origins, key=operator.itemgetter(0))
```

eellison reviewed (approved):
Nit: use ordered set https://github.com/pytorch/pytorch/blob/ee3a7bdc2d0966ab788e48e640879fcf4cd1c69e/torch/_inductor/fx_passes/group_batch_fusion.py#L755

anijain2305 committed:
Update on "[inductor][scheduler] Use set for origin"


xref - https://github.com/pytorch/pytorch/issues/119440

This avoids node > node comparison if the origin order is same in the origins tuple. However, I am unable to come up with a test case where this could happen.

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx peterbell10 ipiszy yf225 chenyang78 kadeng muchulee8 aakhundov ColinPeppler

[ghstack-poisoned]

anijain2305 commented:
@pytorchbot merge

anijain2305 commented:
@pytorchbot merge

Skylion007 commented on torch/_inductor/scheduler.py:
```diff
@@ -2244,9 +2245,13 @@ def get_order(n):
 
-        origins = [(get_order(e), e) for n in node.get_nodes() for e in n.node.origins]
+        # Use a dict to have ordering
+        origins = {
```
Use `dict.fromkeys()` instead here.
